### PR TITLE
[Data format] Serialization for Immutable Graph and HeteroGraph

### PR DESCRIPTION
**Pull request by VoVAllen**
## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

**Comment by VoVAllen** [comment]
@BarclayII I'm not familiar with HeteroGraph enough. Could you give me some suggestions on the `HeteroGraph` part in `test_serialize.cc`? Such as how to construct more HeteroGraphs, verify edge ids and edge directions? Thanks.

**Comment by BarclayII** [comment]
Could we expose the interface to Python (e.g. have `DGLHeteroGraph.save()` and `DGLHeteroGraph.load()`) and do unit test there instead?

**Comment by VoVAllen** [comment]
Sure. I expect these would be used for pickling in the future.

File: src/graph/graph_serializer.h [hunk 1]
```diff
@@ -0,0 +1,25 @@
+/*!
+ *  Copyright (c) 2018 by Contributors
+ * \file graph/graph_serializer.cc
+ * \brief DGL serializer APIs
+ */
+
+#pragma once
```
Comment (jermainewang) [inline src/graph/graph_serializer.h]: Please use the same header guard as the others.

File: src/graph/heterograph.h [hunk 2]
```diff
@@ -170,7 +170,20 @@ class HeteroGraph : public BaseHeteroGraph {
...
+  /*! \return Save HeteroGraph to stream, using CSRMatrix */
+  void Save(dmlc::Stream* fs) const;
+
+
  private:
+  // To create empty class
+  friend class Serializer;
+
+  // Empty Constructor, only for serializer
+  HeteroGraph() : BaseHeteroGraph(static_cast<GraphPtr>(nullptr)) {}
```
Comment (jermainewang) [inline src/graph/heterograph.h]: If you'd like to add a default constructor, please first add one in `BaseHeteroGraph`.

File: src/graph/heterograph.cc [hunk 3]
```diff
@@ -494,6 +500,39 @@ CompactGraphs(const std::vector<HeteroGraphPtr> &graphs) {
...
+  CHECK_EQ(magicNum, kDGLSerialize_HeteroGraph) << "Invalid HeteroGraph Data";
+  auto meta_grptr = new ImmutableGraph(static_cast<COOPtr>(nullptr));
+  CHECK(fs->Read(meta_grptr)) << "Invalid Immutable Graph Data";
+  uint64_t num_relation_graphs;
+  CHECK(fs->Read(&num_relation_graphs)) << "Invalid num of relation graphs";
+  std::vector<HeteroGraphPtr> relgraphs;
+  for (size_t i = 0; i < num_relation_graphs; ++i) {
+    UnitGraph* ugptr = Serializer::EmptyUnitGraph();
+    CHECK(fs->Read(ugptr)) << "Invalid UnitGraph Data";
+    relgraphs.emplace_back(dynamic_cast<BaseHeteroGraph*>(ugptr));
```
Comment (jermainewang) [inline src/graph/heterograph.cc]: There is no need to up-cast a subclass pointer to its parent graph pointer.

Comment (jermainewang) [reply src/graph/heterograph.cc]: Manipulating raw pointers with shared pointer as result is dangerous. Why not the following?
```c++
std::vector<HeteroGraphPtr> relgraphs;
for (...) {
  UnitGraphPtr ugptr = Serializer::EmptyUnitGraph();
  CHECK(fs->Read(ugptr.get())) << ... ;
  relgraphs.push_back(ugptr);
}
```

Comment (jermainewang) [reply src/graph/heterograph.cc]: Also I really suggest implementing serializer handlers such as one [here](https://github.com/dmlc/dgl/blob/master/include/dgl/runtime/serializer.h). This automatically allows serializing type `std::vector<T>` where `T` has specialized `Handler<T>`.

File: src/graph/unit_graph.cc [hunk 4]
```diff
@@ -1152,8 +1155,17 @@ SparseFormat UnitGraph::SelectFormat(SparseFormat preferred_format) const {
     return SparseFormat::COO;
 }
 
+UnitGraph* UnitGraph::EmptyGraph() {
+  auto src = NewIdArray(0);
+  auto dst = NewIdArray(0);
+  auto mg = CreateUnitGraphMetaGraph(1);
+  COOPtr coo(new COO(mg, 0, 0, src, dst));
+  return new UnitGraph(mg, nullptr, nullptr, coo);
+}
```
Comment (jermainewang) [inline src/graph/unit_graph.cc]: Please use shared pointer. Returning raw pointer is dangerous and it's hard to track liveness.

**Comment by VoVAllen** [comment]
Thanks a lot. I'm keeping learning how to write high quality C++ codes.

File: src/graph/heterograph.cc (hunk 3 above)
Comment (VoVAllen) [reply src/graph/heterograph.cc]: > Also I really suggest implementing serializer handlers such as one [here](https://github.com/dmlc/dgl/blob/master/include/dgl/runtime/serializer.h). This automatically allows serializing type `std::vector<T>` where `T` has specialized `Handler<T>`.

Actually it's the same as the current implementation using `Load/Save` with the template class below.
https://github.com/dmlc/dmlc-core/blob/649be18a8c55c48517861d67158a45dec54992ee/include/dmlc/serializer.h#L104

**Pull request merged**